mvlearn/decomposition/groupica.py

```diff
@@ -0,0 +1,326 @@
+# Copyright 2019 NeuroData (http://neurodata.io)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# Authors: Pierre Ablin, Hugo Richard
+
+import numpy as np
+from picard import picard
+import scipy.linalg as linalg
+from sklearn.decomposition import fastica
+from sklearn.utils.validation import check_is_fitted
+
+
+from ..utils.utils import check_Xs
+from .base import BaseDecomposer
+from .grouppca import GroupPCA
+
+
+class GroupICA(BaseDecomposer):
+    r"""
+    Group Independent component analysis.
+    Each dataset in `Xs` is reduced with usual PCA (this step is optional).
+    Then, datasets are concatenated in the features direction, and a PCA is
+    performed on this matrix, yielding a single dataset. ICA is finally
+    performed yielding the output dataset S. The unmixing matrix W
+    corresponding to data X are obtained by solving
+    argmin_{W} ||S - WX||^2.
+
+    Parameters
+    ----------
+    n_components : int, optional
+        Number of components to extract. If None, n_components is set to
+        the minimum number of features in the dataset
+
+    n_individual_components : int or list of int or 'auto', optional
+        The number of individual components to extract as a preprocessing.
+        If None, no preprocessing is applied. If an `int`, each dataset
+        is reduced to this dimension. If a list, the dataset `i` is
+        reduced to the dimension `n_individual_components[i]`.
+        If `'auto'`, set to the minimum between n_components and the
+        smallest number of features in each dataset.
+
+    multiple_outputs : bool, optional
+        If True, the `.transform` method returns one dataset per view.
+        Otherwise, it returns one dataset, of shape (n_samples, n_components)
```

Comment:
this needs to be discussed as I feel it will be very common. How about when you have a pipeline in mvlearn?

Reply:
Indeed it's an important question. I think that it is ok to have a single dataset as output, because is allows for seamless integration with sklearn. To me, the most important thing is that the user is well aware of this.

Reply:
to integrate with sklearn you can have ConcatTransformer that you insert in a pipeline?

Reply:
Yes I think this is something that definitely should be implemented.
I'm not sure it is in the library yet, any idea @rflperry ?

Anyways, if `multiple_outputs=False`, we can already integrate this in a sklearn pipeline:



```
import numpy as np

from sklearn.decomposition import PCA
from sklearn.pipeline import Pipeline
from mvlearn.decomposition import GroupICA


pipe = Pipeline([('groupica', GroupICA()),
                 ('pca', PCA())])

pipe.fit_transform(np.random.randn(2, 10, 3))

```

Reply:
We have no pipeline module in the library but I am fully in support of adding it as it seems useful.

I see that `multiple_output=False` is the default. However, there will be decomposition algorithms which potentially can only output multiple outputs and thus `multiple_outputs=True`. I think it would be confusing to have some return multiple, and some return just one by default. My stance is to have `True` as default unless every algorithm can output a single view by default.

Reply:
- I think that the sklearn pipeline works out of the shelf in mvlearn, as there is no check that the output of a .transform is a 2d array in `sklearn.pipeline.Pipeline`. For instance, 

```
import numpy as np

from sklearn.decomposition import PCA
from sklearn.pipeline import Pipeline
from mvlearn.decomposition import GroupICA


pipe = Pipeline([('groupica1', GroupICA(multiple_outputs=True)),
                          ('groupica2',GroupICA())])

pipe.fit_transform(np.random.randn(2, 10, 3))
```
work perfectly fine. To me, @agramfort 's comment was more about transforming multiview datasets into singleview datasets, so that for instance it is painless to do a ridge regression on the stacked features, just by using a pipeline.
I will give the PR a try.

- I share your concern. I make multiple_output=False the default for both GroupPCA and GroupICA because it is the most natural: both methods work by merging the datasets in a single one. Obtaining multiple outputs requires to perform some weird least-squares estimation, which is not really natural to me. Also, it is not obvious how to recover the singles output from the multiple outputs (e.g. averaging does not work)

It would make things more consistent to always output multiple views by default, but it comes at the price of what is natural for some methods.

Reply:
The unnatural least squares thing is the computation of the individual unmixing matrices yes? I do see in the group ICA paper they back-construct individual results as well as the group result so that would be a clear multiview output option.

It seems to me that in the generalized case, a decomposition algorithm will compute some combination of individual and common results. I see two ways to return a result:
1) Return the common decomposition result and individual results are left as attributes.
2) Return a list of individual results (whether combined with the common result or not is algo-specific) and leave the common result as an attribute.

In case 2, there is then the issue if an algorithm only computes a common result. One could just return a list of that output repeated since in a way each view is decomposed to the same thing. Of course, that isn't particularly efficient but is a viable solution.

Reply:
@agramfort comment below supports option 1 I do believe. I am satisfied with this as the default `mvlearn.decomposition` api although the transform api in the `mvlearn.embed` module (where all the cca's lie) is a list output. Unless one has an issue with this across-module discrepancy I will update our API documentation and this PR can proceed with a single output.